cookbooks/fb_networkmanager/libraries/resource.rb

```diff
@@ -0,0 +1,92 @@
+#
+# Cookbook:: fb_networkmanager
+# Recipe:: default
+#
+# Copyright (c) 2020-present, Vicarious, Inc.
+# Copyright (c) 2020-present, Facebook, Inc.
+# All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+module FB
+  class Networkmanager
+    module Resource
+      NM_CONN_DIR = '/etc/NetworkManager/system-connections'.freeze
+      PREFIX = 'fb_networkmanager'.freeze
+
+      def conf_path(name, normalize = true)
+        if normalize
+          n = "#{PREFIX}_#{name.downcase.gsub(' ', '_')}"
+        else
+          n = name
+        end
+        ::File.join(NM_CONN_DIR, n)
+      end
+
+      def determine_files(name, config)
+        cfile = conf_path(name)
+        migratefile = fromfile = nil
+        if config['_migrate_from']
+          migratefile = conf_path(config['_migrate_from'], false)
+          # Once we've created a file, we should no longer pay attention to the
+          # old file
+          if ::File.exist?(cfile)
+            fromfile = cfile
+          elsif ::File.exist?(migratefile)
+            Chef::Log.info(
+              "#{PREFIX}: Building #{name} based on #{migratefile}",
+            )
+            fromfile = migratefile
+          else
+            # Neither the new file nor the old file exist...
+            Chef::Log.warn(
+              "#{PREFIX}: We were asked to migrate from #{migratefile} but " +
+              'that does not exist, so making a fresh config',
+            )
+            fromfile = cfile
+          end
+          config.delete('_migrate_from')
+        else
+          fromfile = cfile
+        end
+
+        return {
+          'config' => cfile,
+          'from' => fromfile,
+          'migrate' => migratefile,
+        }
+      end
+
+      def allowed_connections(node)
+        node['fb_networkmanager']['system_connections'].keys.map do |x|
+          "#{PREFIX}_#{x.downcase.gsub(' ', '_')}"
+        end
+      end
+
+      # the magic here is all DeepMerge and testing this is arguably a bit
+      # silly, but since this is the most important part of the whole cookbook,
+      # we add a test just to make sure no one ever reverses the arguments or
+      # something weird
+      def generate_config_hashes(from_file, desired_config)
+        current = {}
+        if ::File.exist?(from_file)
+          current = IniParse.parse(::File.read(from_file)).to_hash
+        end
+        new_config = Chef::Mixin::DeepMerge.merge(current, desired_config)
```

Comment:
Since you're targeting client machines here, there are cases where it's reasonable to keep what's in the config file (the user may have changed it) and cases where we want the Chef-defined value to be used no matter what. For example, on some connections I may be fine with the user changing the `permissions` field, but on others I want to enforce it's set as I dictate. Similarly for `autoconnect-priority`, I may want to ensure this connection comes first but users can modify the priority order of other connections.

There should be some way to indicate on a per-connection basis which fields (in which sections) I want the Chef value to always be used for and the normal merge can happen on the rest of the fields.

Reply:
There is. Set the value in Chef if you want Chef to win, don't set the value in Chef if you want the user to be able to win.

Reply:
Is there a usecase for having Chef set a default here, but then letting the user override it?

Reply:
I want to be able to provide a default value when the NM default isn't optimal for the setup, so to do that I need a way to say "these settings are mandatory for this connection, these are defaults the user can override".

Reply:
As I mentioned in the Description, I also considered writing in this way, but could come up with no such example. Can you give me an example? I can certainly implement it.

Reply:
One that I actually deal with: I define 3 connection profiles for my users: `corporate`, `guest`, and `shuttle`. I want `corporate` to always be the highest priority, so I want to enforce that `connection.autoconnect-priority` is set to `999` (which seems to be the highest value NM consistently likes, sometimes higher values work and sometimes NM doesn't like them). The `guest` and `shuttle` networks I'd like to be more preferred than other networks, so I set `connection.autoconnect-priority` to `100` as a default value but I don't mind if some users want to make `shuttle` more preferred than `guest`.

Another example I've run into: I may set `802-1x.private-key-password-flags` to `1` (agent-owned, NetworkManager queries the Secret Service DBus API) by default but for some set of users not easily grouped (otherwise I could just do this in the Chef recipe) that's not working at all and I want to allow them to set `802-1x.private-key-password-flags` to `0` (NM-owned, the PSK is stored in `802-1x.private-key-password` in the connection settings file itself).

Reply:
OK, fair enough, I'll think on this.

Reply:
ok @jgoguen I've added a defaults feature. I need to add docs, but feel free to look at the code

Reply:
and docs added.